src/runtime/set.cpp

```diff
@@ -118,22 +118,29 @@ BoxedSet* makeNewSet(BoxedClass* cls, Box* container) {
     assert(isSubclass(cls, frozenset_cls) || isSubclass(cls, set_cls));
 
     BoxedSet* rtn = new (cls) BoxedSet();
-
-    if (container) {
-        if (PyAnySet_Check(container)) {
-            for (auto&& elt : ((BoxedSet*)container)->s) {
-                rtn->s.insert(incref(elt));
-            }
-        } else if (PyDict_CheckExact(container)) {
-            for (auto&& elt : ((BoxedDict*)container)->d) {
-                rtn->s.insert(incref(elt.first));
-            }
-        } else {
-            for (auto elt : container->pyElements()) {
-                _setAddStolen(rtn, elt);
+    try {
+        if (container) {
+            if (PyAnySet_Check(container)) {
+                for (auto&& elt : ((BoxedSet*)container)->s) {
+                    rtn->s.insert(incref(elt));
+                }
+            } else if (PyDict_CheckExact(container)) {
+                for (auto&& elt : ((BoxedDict*)container)->d) {
+                    rtn->s.insert(incref(elt.first));
+                }
+            } else {
+                for (auto elt : container->pyElements()) {
+                    AUTO_DECREF(elt);
```

Comment:
the problem is not that ```_setAddStolen``` is throwing but the implicit call of the ```BoxAndHash``` constructor (inside the ```PyHasher()``` call) is so the exception happens before we are inside ```_setAddStolen```. Maybe I should create a overloaded ```_setAddStolen(BoxedSet* self, STOLEN(Box*) val)``` which handles this case?
In general it's quite dangerous that ```BoxAndHash(box)``` can throw while we are implicit calling it in a lot of places :-(.